src/ui/__tests__/ui_app.js

```diff
@@ -0,0 +1,253 @@
+          id: 'uiapp-test-defaults',
+          title: undefined,
+          order: 0,
+          url: '/app/uiapp-test-defaults',
+          subUrlBase: '/app/uiapp-test-defaults',
+          description: undefined,
+          icon: undefined,
+          linkToLastSubUrl: true,
+          hidden: false,
+          disabled: false,
+          tooltip: ''
+        });
+      });
+
+      it('has a built-in default order of 0', () => {
+        expect(newApp.order).to.be(0);
+      });
+
+      it('has a built-in default template name of ui_app', () => {
+        expect(newApp.templateName).to.be('ui_app');
+      });
+    });
+
+    describe('with spec', () => {
+      const spec = getMockSpec();
+      let newApp;
+      beforeEach(() => {
+        newApp = new UiApp(uiExports, spec);
+      });
+
+      it('copies the ID from the spec', () => {
+        expect(newApp.id).to.be(spec.id);
+      });
+
+      it('copies field values from spec', () => {
+        // test that the fields exist, but have undefined value
+        expect(newApp.main).to.be(spec.main);
+        expect(newApp.title).to.be(spec.title);
+        expect(newApp.description).to.be(spec.description);
+        expect(newApp.icon).to.be(spec.icon);
+        expect(newApp.linkToLastSubUrl).to.be(spec.linkToLastSubUrl);
+        expect(newApp.templateName).to.be(spec.templateName);
+        expect(newApp.order).to.be(spec.order);
+        expect(newApp.navLink).to.eql({
+          id: 'uiapp-test',
+          title: 'UIApp Test',
+          order: 9000,
+          url: '/app/uiapp-test',
+          subUrlBase: '/app/uiapp-test',
+          description: 'Test of UI App Constructor',
+          icon: 'ui_app_test.svg',
+          linkToLastSubUrl: true,
+          hidden: false,
+          disabled: false,
+          tooltip: ''
+        });
+      });
+    });
+
+    describe('reference fields', () => {
+      const spec = getMockSpec({ testSpec: true });
+      let newApp;
+      beforeEach(() => {
+        newApp = new UiApp(uiExports, spec);
+      });
+
+      it('has a reference to the uiExports object', () => {
+        expect(newApp.uiExports).to.be(uiExports);
+      });
+
+      it('has a reference to the original spec', () => {
+        expect(newApp.spec).to.be(spec);
+      });
+
+      it('has a reference to the spec.injectVars function', () => {
+        const helloFunction = () => 'hello';
+        const spec = {
+          id: 'uiapp-test',
+          injectVars: helloFunction
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.getInjectedVars).to.be(helloFunction);
+      });
+    });
+
+    describe('app.getInjectedVars', () => {
+      it('is noop function by default', () => {
+        const spec = {
+          id: 'uiapp-test'
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.getInjectedVars).to.be(noop);
+      });
+    });
+
+    describe('hidden and listed flags', () => {
```

Comment:
I was also a little surprised that these properties have a relationship and the logic previously had no test coverage. But I think if it is hard to figure that out just by looking at the test, would it suffice to just add a comment in the code that all these tests are here to check the logic for the relationship?

Your proposal looks good to me though, and I appreciate you taking the time to present it like that. I have no problem with integrating it. I'll add a comment as well.